pkg/prowgen/podspec.go

```diff
@@ -0,0 +1,502 @@
+				{
+					Name:      "pull-secret",
+					MountPath: "/etc/pull-secret",
+					ReadOnly:  true,
+				},
+				{
+					Name:      "result-aggregator",
+					MountPath: "/etc/report",
+					ReadOnly:  true,
+				},
+				{
+					Name:      "gcs-credentials",
+					MountPath: cioperatorapi.GCSUploadCredentialsSecretMountPath,
+					ReadOnly:  true,
+				},
+			},
+		},
+	},
+	Volumes: []corev1.Volume{
+		{
+			Name: "pull-secret",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{SecretName: "registry-pull-credentials"},
+			},
+		},
+		{
+			Name: "result-aggregator",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{SecretName: "result-aggregator"},
+			},
+		},
+	},
+}
+
+type ciOperatorPodSpecGenerator struct {
+	mutators    []PodSpecMutator
+	buildErrors []error
+}
+
+// PodSpecMutator is a mutation function operating over a PodSpec pointer.
+// The changes performed on the PodSpec must not depend on any previous
+// state of the PodSpec, and must resolve any conflicts it encounters.
+// The mutator function must also correctly handle being called multiple calls
+type PodSpecMutator func(spec *corev1.PodSpec) error
+
+// CiOperatorPodSpecGenerator is a builder-pattern for building PodSpecs that
+// run ci-operator for various purposes, ensuring the consistency of volumes,
+// mounts, parameters and other content.
+type CiOperatorPodSpecGenerator interface {
+	// Add adds one or more mutations to be performed to build the final PodSpec
+	Add(mutators ...PodSpecMutator) CiOperatorPodSpecGenerator
+	// Build returns the PodSpec built by taking the default PodSpec, applying
+	// all added mutators and sorting several list fields by their keys.
+	Build() (*corev1.PodSpec, error)
+}
+
+// NewCiOperatorPodSpecGenerator returns a new CiOperatorPodSpecGenerator instance
+func NewCiOperatorPodSpecGenerator() CiOperatorPodSpecGenerator {
+	return &ciOperatorPodSpecGenerator{}
+}
+
+func aggregateMutator(mutators ...PodSpecMutator) PodSpecMutator {
+	return func(spec *corev1.PodSpec) error {
+		var errs []error
+		for _, f := range mutators {
+			errs = append(errs, f(spec))
+		}
+		return kerrors.NewAggregate(errs)
+	}
+}
+
+func (c *ciOperatorPodSpecGenerator) Add(f ...PodSpecMutator) CiOperatorPodSpecGenerator {
+	c.mutators = append(c.mutators, f...)
+	return c
+}
+
+func addEnvVar(container *corev1.Container, want corev1.EnvVar) error {
+	var found bool
+	for i := range container.Env {
+		if container.Env[i].Name != want.Name {
+			continue
+		}
+		if !reflect.DeepEqual(want, container.Env[i]) {
+			return fmt.Errorf("environment variable '%s' added with different value", want.Name)
+		}
+		found = true
+	}
+	if !found {
+		container.Env = append(container.Env, want)
+	}
+	return nil
+}
+
+func addUniqueParameter(container *corev1.Container, wantArg string) {
+	var foundArg bool
+	for i := range container.Args {
+		if container.Args[i] == wantArg {
+			foundArg = true
```

Comment:
we can `break` the loop here since the arg was found, or even better we can just `return` and the if-not-found check below won't be needed.